### DISCUSSION TITLE
How to do testing?

### DISCUSSION THREAD

**Discussion by ckruse** [maxcountryman/axum-login#168] [Q&A] [answered]
Hi there,

in a FLOSS software I write I'd like to test routes protected by `axum-login`. I use `router.oneshot()` to hit the URL with a test, e.g. like this:

```rust
let app: Router = get_app().await.expect("could not create app");
let mut rq = http::Request::builder().uri("/foo").method(http::Method::GET);
let response = app.oneshot(rq.body(Body::empty()).unwrap()).await.unwrap();
// …
```

What is the suggested approach to run the test with a user logged in?

**Edit:** Sorry, forgot to add the axum-login setup. It boils down to this:

```rust
async fn get_app() -> Router {
    let session_layer = SessionManagerLayer::new(caching_store)
        .with_expiry(Expiry::OnInactivity(
            axum_login::tower_sessions::cookie::time::Duration::hours(1),
        ));

    Router::new()
        .route("/", get(example))
        .layer(AuthManagerLayerBuilder::new(state.clone(), session_layer.clone()).build())
        .layer(session_layer)
        .with_state(state)
}
```

**Comment by maxcountryman** [comment, marked as answer]
In the integration tests [we've used reqwest with cookie management enabled](https://github.com/maxcountryman/axum-login/blob/3435e7f27a0c25588d2bc4ab0ab3177c5e6ddcb8/axum-login/tests/integration-test.rs#L56). If you're using `oneshot`, you could manage the cookies manually but pulling them off the response headers and then attaching them to request headers. It's a little tedious so it's probably worth writing some test helpers for that. It might even make sense to consider incorporating some of those facilities into the crate itself.

**Reply by ckruse** [reply]
Thank you for the pointer. That will work for me.

**Reply by imothee** [reply]
I'm running into the same issue but since we're using Oauth I had to create a standalone route that took a user_id and force logged them in that's only exposed in test. It's not ideal and I'd much rather be able to mock an AuthSession to pass into the Router as a layer so that each test can just inject the right state for the outcome without running a binary and having all the tests somewhat coupled to a server state.

Has there been any followup in another discussion/ticket on where you're thinking about taking the library and what approach to testing you want to bake in?
